### ISSUE TITLE
Warn for functions that should 'obviously' be unsafe

### ISSUE THREAD

**Issue by dimbleby** [rust-lang/rust#27456]
As a noob writing FFI code I have more than once made a mistake something like this

```rust
fn foo(pointer: *const u32) -> u32 {
    unsafe { *pointer + 1 }
}
```

Clearly `foo()` should have been marked unsafe, as a caller can demonstrate by writing the following safe code:

```rust
let ptr = std::ptr::null();
foo(ptr);
```

Wouldn't it be nice if the compiler had warned something along the lines of "foo dereferences one of its pointer arguments, so should be marked as unsafe"?

I recognise that this is only one of a million ways of writing code that should-be-unsafe-but-isn't, and no doubt asking the compiler to warn for all of them would be an impossibly difficult task.

However I'm hoping that this particular case might be fairly low-hanging fruit, and if so I suggest that it would be a useful warning to have.

**Comment by Aatch** [comment]
We already have such an error, it's when you *don't* have an unsafe block in the function. The presence of the unsafe block basically signals that "you know what you're doing". I think that the only cases it would warn in would be obviously-wrong cases like these where the warning would have little value. Otherwise I can imagine it having lots of false-positives.

**Comment by cybergeek94** [comment]
I can see a lint that requires some sort of documentation on unsafe blocks to be useful. It's already a common practice to note why an unsafe block is okay in a comment next to it. What if there was a lint to enforce this? Like `#[deny(missing_unsafe_docs)]`

**Comment by steveklabnik** [comment]
Yes, I concur with @Aatch . there is going to be no good way to figure out stuff like this in a general case.

An RFC might be able to expand on _exactly_ which cases this should be dealt with, if any, but it would need to be extremely precisely specified. As such, I think this is too general to be actionable without one.

**Comment by dimbleby** [comment]
The proposal was absolutely not general.  I really was suggesting only that the particular case where a function dereferences its pointer arguments might be sufficiently low-hanging fruit to be worth doing.

Of course you are welcome to disagree, or to reject the idea for any of a million reasons.  But I'm sorry to see this closed because of a misunderstanding about what it intended to say.

**Comment by cybergeek94** [comment]
@dimbleby 

The lints integrated into rustc are designed to catch common mistakes that would be hard to find otherwise, like not consuming an iterator or a redundant cast. Unsafe code is inherently uncommon; the user assumes all risks involved wheb writing unsafe code, which includes simple mistakes. 

That being said, perhaps you can pitch this idea to @Manishearth for https://github.com/Manishearth/rust-clippy/, or even implement it yourself! Lints don't have to be developed in-tree to be useful. Then, if your lint turns out to be really useful, you'll have much more momentum behind you to get it pulled into rustc.

**Comment by Manishearth** [comment]
Agree with @cybergeek94, rustc lints aren't for catching such things.


I'm open to the following kind of lint in clippy:

 - A non-unsafe function
 - does not contain any branches (so unconditionally does something)
 - dereferences a raw pointer that was provided in an argument

Anything more general than that can have lots of false positives, since it's common to deref raw pointers after checking stuff.
